File: 067.py
#   067
#   Add Binary
#   2015-12-21
#####################################################
class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        if len(a) < len(b):
            return self.addBinary(b, a)
        aa, i, j, c = list(a), len(a)-1, len(b)-1, 0
        while j >= 0:
            t = int(a[i]) + int(b[j]) + c
            aa[i], c, i, j = str(t%2), t/2, i-1, j-1
        while i >= 0:
            t = int(a[i]) + c
            aa[i], c, i = str(t%2), t/2, i-1
        if c == 1:
            aa.insert(0, '1')
        return "".join(aa)

#####################################################
sol = Solution()

a = "111"
b = "110"
r = sol.addBinary(a, b)
print r

File: leetcode/leetcode/067.cpp
//
//  067.cpp
//  leetcode
//
//  Created by Hang Zhang on 12/21/15.
//  Copyright © 2015 Hilings Studio. All rights reserved.
//

#include <iostream>
//#include <vector>
using namespace std;

class Solution {
public:
    string addBinary(string a, string b) {
        if (a.size() < b.size())
            return addBinary(b, a);
        int i = (int)a.size()-1, j = (int)b.size()-1, c = 0;
        while (j >= 0) {
            int t = a[i]-'0' + b[j]-'0' + c;
            a[i] = '0' + t%2;
            c = t / 2;
            i--;
            j--;
        }
        while (i >= 0) {
            int t = a[i]-'0' + c;
            a[i] = '0' + t%2;
            c = t / 2;
            i--;
        }
        if (c == 1)
            a = "1"+a;
        return a;
    }
};

int main(int arg, char *argv[]) {
    // insert code here...
    cout << "LeetCode 067 Add Binary, C++ ...\n\n";
    Solution sol;
    
    string a = "110";
    string b = "111";
    
    string r = sol.addBinary(a, b);
    cout << r;
    
    return 0;
}